### PR TITLE
Adding `metrics/graphite`, `metrics/provider` and `util/conn` packages plus `Emitter` types

### PR DESCRIPTION
**Pull request by jprobinson**
This PR is hoping to knock out both issues #251 and #262.

Included is the the new `Emitter` types for the `metrics/statsd` and `metrics/graphite` packages, both of which utilize the new `util/conn.Manager` connection manager.

The new `metrics/provider.Provider` is a thin layer over the new `Emitter` types along with very basic implementations for `expvar` and `prometheus`.

This PR currently lacks implementations for `dogstatsd` as I wanted to wait for feedback on the `statsd.Emitter` before continuing down the route and creating a `dogstatd.Emitter`.

If this PR is accepted, we can discard PR #259.

### CODE REVIEW

File: metrics/provider/providers.go [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+
+type prometheusProvider struct {
+	namespace string
+	subsystem string
+}
+
+var _ Provider = prometheusProvider{}
+
+// NewPrometheusProvider returns a Prometheus provider that uses the provided
+// namespace and subsystem for all metrics. Help is not provided, which is very
```
Comment (jprobinson) [inline metrics/provider/providers.go]: Do we still need this note since we've added the "help" parameter to all of the implementations?

Comment (peterbourgon) [inline metrics/provider/providers.go]: Great point. Yes, my oversight. Please fix, if you don't mind :)

**Comment by peterbourgon** [comment]
Is this ready for review?

**Comment by jprobinson** [comment]
Yessir!

File: metrics/graphite/graphite.go [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/codahale/hdrhistogram"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/metrics"
+)
+
+// Newcounter will return a metrics.counter with the given name and a base
```
Comment (peterbourgon) [inline metrics/graphite/graphite.go]: newCounter

File: metrics/graphite/graphite.go [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+// With currently ignores fields.
+func (c *counter) With(metrics.Field) metrics.Counter { return c }
+
+func (c *counter) Add(delta uint64) { atomic.AddUint64(&c.count, delta) }
+
+func (c *counter) get() uint64 { return atomic.LoadUint64(&c.count) }
+
+// flush will emit the current counter value in the Graphite plaintext
+// protocol to the given io.Writer.
+func (c *counter) flush(conn io.Writer, prefix string) {
```
Comment (peterbourgon) [inline metrics/graphite/graphite.go]: `w io.Writer`

File: metrics/graphite/emitter.go [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+	}
+
+	err := e.flush(conn)
+	if err != nil {
+		e.logger.Log("during", "flush", "err", err)
+	}
+	e.mgr.Put(err)
+}
+
+func (e *Emitter) flush(conn io.Writer) error {
```
Comment (peterbourgon) [inline metrics/graphite/emitter.go]:
```go
w io.Writer
```

File: metrics/graphite/graphite.go [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+	atomic.StoreUint64(&g.value, math.Float64bits(value))
+}
+
+func (g *gauge) Get() float64 {
+	return math.Float64frombits(atomic.LoadUint64(&g.value))
+}
+
+// Flush will emit the current gauge value in the Graphite plaintext
+// protocol to the given io.Writer.
+func (g *gauge) flush(conn io.Writer, prefix string) {
```
Comment (peterbourgon) [inline metrics/graphite/graphite.go]:
```go
w io.Writer
```

File: metrics/graphite/graphite.go [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+		quantiles = append(quantiles, metrics.Quantile{
+			Quantile: quantile,
+			Value:    int64(gauge.Get()),
+		})
+	}
+	sort.Sort(quantileSlice(quantiles))
+	return buckets, quantiles
+}
+
+func (h *windowedHistogram) flush(conn io.Writer, prefix string) {
```
Comment (peterbourgon) [inline metrics/graphite/graphite.go]:
```go
w io.Writer
```

File: metrics/graphite/graphite.go [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+	"math"
+	"sort"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/codahale/hdrhistogram"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/metrics"
+)
```
Comment (peterbourgon) [inline metrics/graphite/graphite.go]:
```go
import (
    // stdlib

    // third-party

    // go-kit/kit/*
)
```

File: util/conn/manager.go [hunk 8]
```diff
@@ -0,0 +1,105 @@
+package conn
```
Comment (peterbourgon) [inline util/conn/manager.go]: This package needs a README and doc comments, but I'll add that myself afterwards.

File: util/conn/manager.go [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+	"time"
+
+	"github.com/go-kit/kit/log"
+)
+
+// Dialer dials a network and address. net.Dial is a good default Dialer.
+type Dialer func(network, address string) (net.Conn, error)
+
+// time.After is a good default afterFunc.
+type afterFunc func(time.Duration) <-chan time.Time
```
Comment (peterbourgon) [inline util/conn/manager.go]: Oops! This needs to be 

```go
// AfterFunc returns a one-shot timer chan. time.After is a good default AfterFunc.
type AfterFunc func(time.Duration) <-chan time.Time
```

**Comment by peterbourgon** [comment]
Amazing! One or two very quick fixes and then LGTM!

**Comment by jprobinson** [comment]
All cleaned up and ready for your blessing, sir. 🙏

File: util/conn/manager.go [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+	"net"
+	"time"
+
+	"github.com/go-kit/kit/log"
+)
+
+// Dialer dials a network and address. net.Dial is a good default Dialer.
+type Dialer func(network, address string) (net.Conn, error)
+
+// time.After is a good default afterFunc.
```
Comment (peterbourgon) [inline util/conn/manager.go]: Doc comment needs to be of the form `AfterFunc ...`

**Comment by peterbourgon** [comment]
🎉

**Comment by peterbourgon** [comment]
Whoops, accidentally left for dinner before clicking green. Hooray! Thank you!

**Pull request merged**